kernel/h1: Add support for both flash banks.

osenft opened the pull request:
This change adds support to the H1 flash driver for both flash banks.
In order to be able to use the existing interface / design, we now
need to pass the target address to both `H1bHw.set_transaction()` and
`H1bHw.trigger()`.

Further, this change adds `ReturnCode`s to the `Hardware` trait in
preparation to handle these in the driver code in a subsequent
commit.

osenft committed:
kernel/h1: Add support for both flash banks.

This change adds support to the H1 flash driver for both flash banks.
In order to be able to use the existing interface / design, we now
need to pass the target address to both `H1bHw.set_transaction()` and
`H1bHw.trigger()`.

Further, this change adds `ReturnCode`s to the `Hardware` trait in
preparation to handle these in the driver code in a subsequent
commit.

jrvanwhy commented on kernel/h1/src/hil/flash/h1_hw.rs:
```diff
@@ -265,27 +266,56 @@ impl super::hardware::Hardware for H1bHw {
+        }
+
+        if relative_offset + size > H1_FLASH_BANK_SIZE / BYTES_PER_WORD {
+            return ReturnCode::EINVAL
```
This is a lot of complexity for untestable code. Can some of this logic be moved out of `H1bHw`?

osenft replied:
Good point. I moved that logic out into the `Driver`.

osenft replied:
If you like I can squash the two commits and force push if that makes it easier to read.

osenft committed:
Implement fixes for review comments.

The main change here is to move logic out of flash::H1bHw and instead
explicitly specify the flash bank that we're operating on.

jrvanwhy commented on kernel/h1/src/hil/flash/hardware.rs:
```diff
@@ -14,8 +14,19 @@
+pub enum Bank {
+    Unknown,
+    Zero,
+    One,
```
Why not just:

```rust
pub enum Bank {
    Zero,
    One,
}
```

then use `Option<Bank>` (or a similarly-structured custom enum type) wherever you need a third value.

jrvanwhy replied:
In particular, this would make `H1bHw::trigger` simpler.

osenft replied:
I tried to compromise wrt. required changes. But you're right, removing `Unknown` is absolutely better. Please see latest changes.

dann21 commented on kernel/h1/src/hil/flash/driver.rs:
```diff
@@ -63,13 +63,26 @@ impl<'d, A: Alarm<'d>, H: Hardware> FlashImpl<'d, A, H> {
+    if target < WORDS_PER_BANK {
+        super::hardware::Bank::Zero
+    } else {
+        super::hardware::Bank::One
```
Should we assert `target < 2*WORDS_PER_BANK` here?

osenft replied:
The latest commits handle this correctly.

dann21 replied:
Yep -- tx.

dann21 commented on kernel/h1/src/hil/flash/hardware.rs:
```diff
@@ -28,14 +39,15 @@ pub trait Hardware {
 
     /// Set flash transaction parameters (word offset and size). The word offset
-    /// is relative to the start of flash and the size is one less than the
+    /// is relative to the start of a flash bank and the size is one less than the
```
This `size-1` thing feels like a questionable choice to me.  It seems likely to lead to confusion in the future.

Why?

osenft replied:
The `Hardware` trait is really just a super thin layer above the actual HW registers. And the `-1` is just how it is in the actual hardware.

dann21 replied:
Sigh...OK, can't fix that.

dann21 reviewed (approved):
LGTM, with a couple of comments.

osenft committed:
Implement fixes for review comments.

The main change here is to remove Bank::Invalid and replace it with
the use of Option<Bank> where appropriate. This required some
additional (and previously missing) error handling.

dann21 commented:
Sigh...
OK, can't fix that.

On Thu, May 13, 2021 at 9:11 AM Oskar Senft ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In kernel/h1/src/hil/flash/hardware.rs
> <https://github.com/google/tock-on-titan/pull/241#discussion_r631798054>:
>
> > @@ -28,14 +39,15 @@ pub trait Hardware {
>      fn read_error(&self) -> u16;
>
>      /// Set flash transaction parameters (word offset and size). The word offset
> -    /// is relative to the start of flash and the size is one less than the
> +    /// is relative to the start of a flash bank and the size is one less than the
>
> The Hardware trait is really just a super thin layer above the actual HW
> registers. And the -1 is just how it is in the actual hardware.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/google/tock-on-titan/pull/241#discussion_r631798054>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AECVBXYK457JAL6ZLB5UTGLTNPFYRANCNFSM44YYFN7Q>
> .
>

osenft committed:
Add test coverage for both flash banks and fix fake flash to
correctly handle requests for both banks.

dann21 commented on kernel/h1/src/hil/flash/driver.rs:
```diff
@@ -63,13 +67,36 @@ impl<'d, A: Alarm<'d>, H: Hardware> FlashImpl<'d, A, H> {
         self.smart_program(ERASE_OPCODE, /*max_attempts*/ 45, /*final_pulse_needed*/ false,
-                           /*timeout_nanoseconds*/ 3_353_267,
-                           /*target*/ page * super::WORDS_PER_PAGE, /*size*/ 1);
+                           /*timeout_nanoseconds*/ 3_353_267, self.write_bank.get(),
```
Where the _heck_ did this bizarre-looking magic number come from?

osenft replied:
I assume @jrvanwhy computed that from the H1B specs ...

dann21 replied:
Maybe not yours to dig up, but IMO anything like this ought to include a comment with a pointer to the relevant spec.
